### PR TITLE
Global exception dev

### PR DESCRIPTION
**Pull request by don-bigdad**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -16,8 +16,20 @@
...
         <maven.checkstyle.plugin.configLocation>checkstyle.xml</maven.checkstyle.plugin.configLocation>
+        <lombok.mapstruct.binding.version>0.2.0</lombok.mapstruct.binding.version>
+        <mapstruct.version>1.5.5.Final</mapstruct.version>
     </properties>
     <dependencies>
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-validator</artifactId>
+            <version>8.0.1.Final</version>
+        </dependency>
```
Comment (ystankevych) [inline pom.xml]:
```suggestion
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-validation</artifactId>
        </dependency>
```
you could use spring-boot-starter-validation as it contains all needed dependencies

File: src/main/java/com/example/springbootbookshop/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.example.springbootbookshop.dto;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import java.math.BigDecimal;
+
+public record CreateBookRequestDto(@NotNull String title, @NotNull String author,
```
Comment (ystankevych) [inline src/main/java/com/example/springbootbookshop/dto/CreateBookRequestDto.java]: Maybe NotBlank will be more suitable for String validation. As NotNull not allow only null values, but field can be empty. As for NotBlank -> a String field constrained with @NotBlank must be not null, and the trimmed length must be greater than zero

**Review by ystankevych — APPROVED** [review]
Hi, Great job. I approved but left some comments.

File: pom.xml (hunk 1 above)
Comment (NazarSmith228) [reply pom.xml]: agree with @ystankevych - it is better to use `spring-boot-starter-validation`

File: src/main/java/com/example/springbootbookshop/controller/BookController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    @DeleteMapping(value = "/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteBookById(@PathVariable Long id) {
+        bookService.deleteById(id);
+    }
+
+    @GetMapping(value = "/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public BookDto getBookById(@PathVariable Long id) {
```
Comment (NazarSmith228) [inline src/main/java/com/example/springbootbookshop/controller/BookController.java]: just FYI

you may also validate such things as @PathVariable, @RequestParams etc. It looks like:
```
@PathVariable @Positive Long id
```

In order to be able to perform such validation you need to annotate your controller class with @Validated annotation

again, this is just FYI - apply it or not :)

File: src/main/java/com/example/springbootbookshop/dto/CreateBookRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.example.springbootbookshop.dto;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotBlank;
+import java.math.BigDecimal;
+
+public record CreateBookRequestDto(@NotBlank String title,
+                                   @NotBlank String author,
+                                   @NotBlank String isbn,
+                                   @NotBlank @Min(0) BigDecimal price,
```
Comment (NazarSmith228) [inline src/main/java/com/example/springbootbookshop/dto/CreateBookRequestDto.java]: `@Min(0)` looks weird :)

could you use `@Positive` or `@PositiveOrZero`?

File: src/main/java/com/example/springbootbookshop/exception/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+@ControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers, HttpStatusCode status,
+            WebRequest request) {
+        Map<String,Object> body = new LinkedHashMap<>();
+        body.put("timestampt", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
```
Comment (NazarSmith228) [inline src/main/java/com/example/springbootbookshop/exception/CustomGlobalExceptionHandler.java]: time is pretty useless in response body - it does not bring any value to the client

status is obvious as well - no sense to have it in body

File: src/main/java/com/example/springbootbookshop/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
```
Comment (NazarSmith228) [inline src/main/java/com/example/springbootbookshop/exception/CustomGlobalExceptionHandler.java:21]: do you have any other exceptions except for `MethodArgumentNotValidException`?

does it make sense to handle them as well?

File: src/main/java/com/example/springbootbookshop/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -1,28 +1,58 @@
...
+    public BookDto update(CreateBookRequestDto bookRequestDto, Long id) {
+        Optional<Book> bookOptional = bookRepository.findById(id);
+        if (bookOptional.isPresent()) {
+            Book book = bookMapper.toBook(bookRequestDto);
+            book.setId(id);
+            bookRepository.save(book);
+            return bookMapper.toDto(book);
+        } else {
+            throw new EntityNotFoundException("Book with ID " + id + " not found");
+        }
```
Comment (NazarSmith228) [inline src/main/java/com/example/springbootbookshop/service/impl/BookServiceImpl.java]: let's improve this a bit :)

first thing is that you can leverage Mapstruct to perform the update:
```
void updateBook(CreateBookRequestDto dto, @MappingTarget Book book);
```

then:
```
Book bookToUpdate = bookRepository.findById(id).orElseThrow(...);
bookMapper.updateBook(dto, bookToUpdate);
bookRepository.save(book);
return bookMapper.toDto(book);
```

File: src/main/java/com/example/springbootbookshop/service/BookService.java [hunk 8]
```diff
@@ -1,10 +1,18 @@
...
+    Book save(CreateBookRequestDto book);
 
-    List<Book> findAll();
+    List<BookDto> findAll();
+
+    BookDto getBookById(Long id);
+
+    void deleteById(Long id);
+
+    public BookDto update(CreateBookRequestDto bookRequestDto, Long id);
```
Comment (NazarSmith228) [inline src/main/java/com/example/springbootbookshop/service/BookService.java]: `public` is redundant here

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: pom.xml [hunk 9]
```diff
@@ -16,8 +16,15 @@
...
         <java.version>17</java.version>
         <maven.checkstyle.plugin.configLocation>checkstyle.xml</maven.checkstyle.plugin.configLocation>
+        <lombok.mapstruct.binding.version>0.2.0</lombok.mapstruct.binding.version>
+        <mapstruct.version>1.5.5.Final</mapstruct.version>
     </properties>
     <dependencies>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-validation</artifactId>
+            <version>3.1.4</version>
```
Comment (NazarSmith228) [inline pom.xml:26]: there is no need to specify the version here

**Review by NazarSmith228 — APPROVED** [review]
LGTM

**Pull request merged**